Fixes Part of #8472: Migrate Exploration data service

srijanreddy98 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #8472.
2. This PR does the following: Migrates Exploration data service

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

srijanreddy98 committed:
Migrate editable exploration backend api service

srijanreddy98 committed:
Fix types

srijanreddy98 committed:
Migrate exploration data service

srijanreddy98 committed:
Refactor to backend api service

srijanreddy98 committed:
Refactor to backend api service

srijanreddy98 committed:
Refactor to backend api service

srijanreddy98 committed:
Refactor to backend api service

srijanreddy98 commented:
@kevintab95 @nithusha21 PTAL!

srijanreddy98 committed:
Fix

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into eds

srijanreddy98 committed:
Add wait for element to be clickable

kevintab95 commented on core/templates/pages/exploration-editor-page/services/exploration-data.service.ts:
```diff
@@ -17,187 +17,212 @@
-                  }
-                }
-              }
-              $rootScope.$applyAsync();
```
Can you add this in the appropriate place in the component file that uses this method? Ditto below.

srijanreddy98 replied:
I already did that in #11558.

kevintab95 replied:
ah okay, I missed that. Thanks for confirming!

kevintab95 commented on core/templates/pages/exploration-editor-page/services/exploration-data.service.ts:
```diff
@@ -17,187 +17,212 @@
+        }
       }
-    };
-    $rootScope.$applyAsync();
```
ditto

srijanreddy98 replied:
Done! Added it in exploration save service.

kevintab95 reviewed (changes requested):
Looks great @srijanreddy98! Thanks! Just 1 comment. PTAL.

srijanreddy98 committed:
Add a couple more applyAsync

srijanreddy98 commented:
@kevintab95 addressed your comments, PTAL!

kevintab95 reviewed (approved):
LGTM, thanks @srijanreddy98!

U8NWXD commented on core/tests/protractor_utils/workflow.js:
```diff
@@ -131,8 +131,11 @@ var createExplorationAsAdmin = async function() {
-  var prePublicationButtonElem = element(by.css(
+  const prePublicationButtonElem = element(by.css(
     '.protractor-test-confirm-pre-publication'));
+  await waitFor.elementToBeClickable(
+    prePublicationButtonElem,
+    'Save changes button taking too long to be clickable');
   await prePublicationButtonElem.isPresent();
   await prePublicationButtonElem.click();
```
You can use `action.click` from `protractor_utils/action.js` to do all of this in one line

U8NWXD reviewed (approved):
LGTM. I just left one note about a helper function you can use if you like

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into ebas

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into eds

srijanreddy98 committed:
Merge branch 'ebas' into eds

srijanreddy98 committed:
Migrate exploration-backend-api service

srijanreddy98 committed:
Migrate exploration-backend-api service

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into eds

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into eds

srijanreddy98 committed:
Add .

srijanreddy98 committed:
Add .

srijanreddy98 committed:
Update workflow.js

srijanreddy98 committed:
Fix exploration diff service

srijanreddy98 committed:
Update workflow.js

kevintab95 commented:
Removing LGTM label because we still need to fix the e2e failures here before it can be merged.

darksun27 committed:
e2e_Patch 1.1: Fix creator_learner_dashboard_suite
- added waitForAngularEnabled
- added waitForElementToBeClickable
- Fixed null index problem

darksun27 committed:
e2e_Patch 1.1.1 Fix Minor Linting Issues

srijanreddy98 committed:
Merge branch 'develop' into eds

darksun27 committed:
Merge branch 'eds' of https://github.com/srijanreddy98/oppia into eds

Commit:
remove waitForAngularEnabled from ExplorationPlayerPage.rateExploration

Commit:
Merge remote-tracking branch 'upstream/develop' into HEAD

darksun27 committed:
fixed clickable button

darksun27 committed:
Merge remote-tracking branch 'srijan/eds' into eds

srijanreddy98 committed:
Remove wait for angular calls

Commit:
fix saving flow issues (also fixes the infamous 'Lost Changes' e2e flake)

Commit:
Merge remote-tracking branch 'upstream/develop' into HEAD

Commit:
prevent changes when autosave is happening

Commit:
fix category input not visible error

srijanreddy98 committed:
Revert of last two commits (relative to this commit)

srijanreddy98 committed:
Auto save indicator

srijanreddy98 committed:
Auto save indicator

srijanreddy98 committed:
Auto save indicator

srijanreddy98 committed:
Unit test changes

srijanreddy98 committed:
Fix e2e tests

Commit:
waitFor autosave invisibility before any action on the settings page

Commit:
move waitForAutosave to action.js

Commit:
add some backend logs to debug 'lost changes'

Commit:
print change dicts

Commit:
fix e2e tests

Commit:
club together changes that occur in quick sequence (200ms) into a single autosave draft backend call

Commit:
remove unnecessary logging; add few more action.waitForAutosave()

Commit:
revert changes to exp_services.py

Commit:
do waitForAutosave before navigating to the main tab in the exp editor

seanlip commented on core/templates/pages/exploration-editor-page/services/exploration-data.service.ts:
```diff
@@ -17,186 +17,212 @@
-      return {};
+      this.loggerService.error(
+        'Unexpected call to ExplorationDataService for pathname: ' + pathname);
+      this.autosaveChangeList = undefined;
```
Hi @srijanreddy98 -- I suspect that these lines might be the cause of a bunch of common flakes we are getting in the frontend tests along the lines of "TypeError: this.explorationDataService.autosaveChangeListAsync is not a function".

Could you please provide some insight on why they were added? Thanks!

/cc @U8NWXD 

Example: https://github.com/oppia/oppia/runs/3576133064?check_suite_focus=true

srijanreddy98 replied:
https://github.com/oppia/oppia/pull/11564/files#r706660354

seanlip replied:
Ah, I see -- the returned dict is the list of functions? Got it, thanks.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/exploration-data.service.ts:
```diff
@@ -17,186 +17,212 @@
-        'Unexpected call to ExplorationDataService for pathname ',
-        pathname);
-      // Note: if we do not return anything, Karma unit tests fail.
-      return {};
```
@seanlip This was the behaviour before. I just kept the behaviour as is.
